### PR TITLE
Add support for Queue Consumer concurrency

### PR DESCRIPTION
**Pull request by jbwcloudflare**
What this PR solves / how to test:

This PR adds support for new Queues Consumer configuration options: concurrency_enabled and max_concurrency.

Associated docs issues/PR: In progress

Author has included the following, where applicable:

- [x] Tests
- [x] Changeset

Reviewer has performed the following, where applicable:

- [ ] Checked for inclusion of relevant tests
- [ ] Checked for inclusion of a relevant changeset
- [ ] Checked for creation of associated docs updates
- [ ] Manually pulled down the changes and spot-tested

### CODE REVIEW

**Comment by jbwcloudflare** [comment]
@elithrar LMK if you have any thought here

File: packages/wrangler/src/__tests__/queues.test.ts [hunk 1]
```diff
@@ -348,10 +348,12 @@ describe("wrangler", () => {
 				  -v, --version                   Show version number  [boolean]
 
 				Options:
-				      --batch-size         Maximum number of messages per batch  [number]
-				      --batch-timeout      Maximum number of seconds to wait to fill a batch with messages  [number]
-				      --message-retries    Maximum number of retries for each message  [number]
-				      --dead-letter-queue  Queue to send messages that failed to be consumed  [string]"
+				      --batch-size           Maximum number of messages per batch  [number]
```
Comment (elithrar) [inline packages/wrangler/src/__tests__/queues.test.ts:351]: LGTM

File: packages/wrangler/src/config/environment.ts [hunk 2]
```diff
@@ -374,6 +374,12 @@ interface EnvironmentNonInheritable {
 
 			/** The queue to send messages that failed to be consumed. */
 			dead_letter_queue?: string;
+
+			/** Whether the Queue broker will make concurrent consumer invocations */
```
Comment (elithrar) [inline packages/wrangler/src/config/environment.ts:378]:
```suggestion
			/** Allow the queue consumer to automatically scale out as the backlog grows  */
```

File: packages/wrangler/src/config/environment.ts [hunk 3]
```diff
@@ -374,6 +374,12 @@ interface EnvironmentNonInheritable {
...
+			concurrency_enabled?: boolean;
+
+			/** The maximum number of concurrent consumer Worker executions */
```
Comment (elithrar) [inline packages/wrangler/src/config/environment.ts:381]:
```suggestion
			/** The maximum number of concurrent consumer Worker invocations. Leaving this unset will allow your consumer to scale to the maximum concurrency needed to keep up with the message backlog. */
```

File: packages/wrangler/src/config/environment.ts [hunk 4]
```diff
@@ -374,6 +374,12 @@ interface EnvironmentNonInheritable {
...
+			max_concurrency?: number;
```
Comment (elithrar) [inline packages/wrangler/src/config/environment.ts:382]: What happens if you put a number larger than we support? e.g. 21? Or 0, or -1 - ?

Comment (jbwcloudflare) [reply packages/wrangler/src/config/environment.ts:382]: It will try to send the invalid value to the API, and fail (with a reasonable error message). That is in line with the other settings at the moment. 

I suppose we could add more client side validation to all of the queue settings, but it is a little annoying to maintain them in two places.

Comment (elithrar) [reply packages/wrangler/src/config/environment.ts:382]: I don't think we need more client validation, but:

- What does 0 do? Is it invalid or coerced to null?
- Is -1 in invalid?
- What is the current max?

(These will go into the docs)

**Comment by jbw1991** [comment]
@cloudflare/wrangler Hey folks, could someone start the builds on this please? 🙏

File: packages/wrangler/src/config/environment.ts (hunk 4 above)
Comment (jbw1991) [reply packages/wrangler/src/config/environment.ts:382]: Ah right, so:
* 0 is invalid
* -1 is invalid
* 20 is the current max enforced by the backend. Shall we drop it to 5?

So to summarize, the user's value must be between 1 and the max, inclusive.
They can also omit the value / specify `null` to indicate that they don't care to enforce a limit (we'll use the highest possible value behind the scenes)

Comment (elithrar) [reply packages/wrangler/src/config/environment.ts:382]: Meet in the middle at 10 for now?

Rest makes sense - thanks!

File: .changeset/wild-pandas-destroy.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
+"wrangler": minor
+---
+
+Add support for Queue Consumer concurrency
```
Comment (penalosa) [inline .changeset/wild-pandas-destroy.md:5]: This should have some more detail around the the user facing changes (new config options?)

File: packages/wrangler/src/queues/cli/commands/consumer/add.ts [hunk 6]
```diff
@@ -37,6 +37,15 @@ export function options(yargs: CommonYargsArgv) {
 				type: "string",
 				describe: "Queue to send messages that failed to be consumed",
 			},
+			"concurrency-enabled": {
```
Comment (penalosa) [inline packages/wrangler/src/queues/cli/commands/consumer/add.ts:40]: Why is this called `concurrency-enabled`? If it's a boolean, could it just be `--concurrency`?

Although actually, what about a single flag `--concurrency` an optional numeric argument?

File: packages/wrangler/src/queues/cli/commands/consumer/add.ts [hunk 7]
```diff
@@ -55,6 +64,8 @@ export async function handler(
 			max_wait_time_ms: args.batchTimeout // API expects milliseconds
 				? 1000 * args.batchTimeout
 				: undefined,
+			concurrency_enabled: args.concurrencyEnabled ?? true,
```
Comment (penalosa) [inline packages/wrangler/src/queues/cli/commands/consumer/add.ts:67]: Was concurrency previously defaulted to false/not supported?

File: packages/wrangler/src/queues/cli/commands/consumer/add.ts (hunk 6 above)
Comment (jbwcloudflare) [reply packages/wrangler/src/queues/cli/commands/consumer/add.ts:40]: I suppose we could call it `concurrency` /shrug.

There are currently two options to support the following:
* The user can decide whether they want concurrency turned on or off entirely
* If concurrency is turned on, the user can *optionally* specify a maximum concurrency level (or omit it / pass null to indicate that they do not care to enforce a maximum level).

File: packages/wrangler/src/queues/cli/commands/consumer/add.ts (hunk 7 above)
Comment (jbwcloudflare) [reply packages/wrangler/src/queues/cli/commands/consumer/add.ts:67]: It was previously not supported at all, and we want it to be enabled by default going forward (with the ability to opt out).

Comment (elithrar) [reply packages/wrangler/src/queues/cli/commands/consumer/add.ts:67]: Concurrency is a new feature. We will default to true when this is not explicitly set to `false`.

File: packages/wrangler/src/queues/cli/commands/consumer/add.ts (hunk 6 above)
Comment (elithrar) [reply packages/wrangler/src/queues/cli/commands/consumer/add.ts:40]: To add to what Josh described: our goal is to have `concurrency_enabled` be the default, and to (ideally) have only users that need to strictly limit set `false` or `max_concurrency: $SOME_FIXED_NUMBER`.

Over time, as we allow more concurrent invocations, developers will just benefit from those improvements without having to touch their config.

Comment (elithrar) [reply packages/wrangler/src/queues/cli/commands/consumer/add.ts:40]: If you merge the two properties, how do you express "enable concurrency, and let Cloudflare define the maximum" (which will grow over time) - e.g. effectively an `auto` mode. 

In this proposal, that's just `concurrency_enabled: true` (or `null` / omit it). `max_concurrency` is a control for users who have upstream concerns.

"Technically" this could be a union type but that's not easily represented in TOML nor is it familiar to most users.

File: packages/wrangler/src/queues/cli/commands/consumer/add.ts (hunk 7 above)
Comment (jbwcloudflare) [reply packages/wrangler/src/queues/cli/commands/consumer/add.ts:67]: @elithrar I'm actually second guessing this now, sorry for the back and forth.

I was trying to be a bit forward looking by defaulting to `true` here in the CLI command. But we probably just want the CLI behavior to match the API behavior (default to `false`) and then both will be "cleaned up" at the same time when we decide to flip the default behavior for a missing value.

File: .changeset/wild-pandas-destroy.md (hunk 5 above)
Comment (jbwcloudflare) [reply .changeset/wild-pandas-destroy.md]: ✅ Updated

File: packages/wrangler/src/queues/cli/commands/consumer/add.ts (hunk 7 above)
Comment (elithrar) [reply packages/wrangler/src/queues/cli/commands/consumer/add.ts:67]: How about:

- We give users time - 1 week or so as we're in beta - to set `max_concurrency` if they need to.
- After the deadline, we start respecting concurrency settings and actually scaling up. For most users, `max_concurrency = null` is implied and thus the maximum.

No need for `concurrency_enabled` at all? @jbw1991

Comment (jbwcloudflare) [reply packages/wrangler/src/queues/cli/commands/consumer/add.ts:67]: Yeah, this is a good tradeoff. We can get rid of `concurrency_enabled`.
For the next few weeks, the API will assume that a missing/null `max_concurrency` should fallback to a default value of 1.
After communicating the change and allowing some time, we'll change the default to 'unlimited'.
There will be no way to use 'unlimited' mode until the API change goes live. But this is a fair price to pay for the cleaner API. Users can just select the maximum possible limit in the meantime.

Comment (elithrar) [reply packages/wrangler/src/queues/cli/commands/consumer/add.ts:67]: w.r..t wrangler flags:

* `--max-concurrency=N` to set a cap
* `--no-max-concurrency` to "null out" the setting (passes an explicit null)

Comment (jbwcloudflare) [reply packages/wrangler/src/queues/cli/commands/consumer/add.ts]: We discussed offline, but the `--no-max-concurrency` is more relevant for a `queues consumer update` command, which doesn't exist yet.

Everything else should be good to go. `--max-concurrency=N` is supported

File: packages/wrangler/src/queues/cli/commands/consumer/add.ts (hunk 6 above)
Comment (jbwcloudflare) [reply packages/wrangler/src/queues/cli/commands/consumer/add.ts]: After some internal discussion, we went ahead and consolidated down to a single `max_concurrency` setting.
Behind the scenes, a missing/null `max_concurrency` will fallback to a default value of 1, temporarily, for the next few weeks.
After communicating the change, the fallback/default will change to 'unlimited' mode.

**Comment by jbwcloudflare** [comment]
@OilyLime Hey just FYI, I finished up support for the concurrency settings over here on my fork.

File: packages/wrangler/src/config/validation.ts [hunk 8]
```diff
@@ -2230,12 +2231,13 @@ const validateConsumer: ValidatorFn = (diagnostics, field, value, _config) => {
 
 	const options: {
 		key: string;
-		type: "number" | "string";
```
Comment (OilyLime) [inline packages/wrangler/src/config/validation.ts:2233]: I think you can revert this change. I made it originally because I had to support `concurrency_enabled`.

**Review by OilyLime — APPROVED** [review]

File: packages/wrangler/src/config/validation.ts (hunk 8 above)
Comment (jbwcloudflare) [reply packages/wrangler/src/config/validation.ts:2233]: Good call, but I'll just leave it, doesn't hurt to have support for booleans here.

File: .changeset/wild-pandas-destroy.md [hunk 9]
```diff
@@ -0,0 +1,8 @@
+---
+"wrangler": minor
+---
+
+feature: add support for Queue Consumer concurrency
+
+Consumer concurrency allows a consumer Worker processing messages from a queue to automatically scale out horizontally in order to keep up with the rate that messages are being written to a queue.
+The new `max_concurrency` setting can be used to limit the maximum number of concurrent consumer Worker invocations.
```
Comment (petebacondarwin) [inline .changeset/wild-pandas-destroy.md:8]: Out if curiosity, how does queue concurrency relate to queue batches?
If you set the concurrency to 1, will you just receive larger batches in that single invocation?

File: packages/wrangler/src/__tests__/publish.test.ts [hunk 10]
```diff
@@ -7386,6 +7386,46 @@ export default{
...
+			"Total Upload: xx KiB / gzip: xx KiB
+			Uploaded test-name (TIMINGS)
+			Published test-name (TIMINGS)
+			  https://test-name.test-sub-domain.workers.dev
+			  Consumer for queue1
+			Current Deployment ID: Galaxy-Class"
+		`);
+		});
+
+		it("should support queue consumer concurrency with a max concurrency specified", async () => {
```
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/publish.test.ts:7403]: Can we have a test where `null` is passed for `max_concurrency`?

File: packages/wrangler/src/__tests__/queues.test.ts [hunk 11]
```diff
@@ -382,13 +384,14 @@ describe("wrangler", () => {
...
 							max_wait_time_ms: 10 * 1000,
+							max_concurrency: 3,
 						},
 						dead_letter_queue: "myDLQ",
 					};
 					mockPostRequest("testQueue", expectedBody);
 
 					await runWrangler(
-						"queues consumer add testQueue testScript --env myEnv --batch-size 20 --batch-timeout 10 --message-retries 3 --dead-letter-queue myDLQ"
+						"queues consumer add testQueue testScript --env myEnv --batch-size 20 --batch-timeout 10 --message-retries 3 --max-concurrency 3 --dead-letter-queue myDLQ"
```
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/queues.test.ts:394]: Can we have a test that uses `--no-max-concurrency`?

File: packages/wrangler/src/queues/cli/commands/consumer/add.ts [hunk 12]
```diff
@@ -37,6 +37,11 @@ export function options(yargs: CommonYargsArgv) {
 				type: "string",
 				describe: "Queue to send messages that failed to be consumed",
 			},
+			"max-concurrency": {
+				type: "number",
+				describe:
+					"The maximum number of concurrent consumer Worker invocations",
```
Comment (petebacondarwin) [inline packages/wrangler/src/queues/cli/commands/consumer/add.ts:43]: Might be worth adding that only positive integers are valid values?

File: .changeset/wild-pandas-destroy.md (hunk 9 above)
Comment (jbwcloudflare) [reply .changeset/wild-pandas-destroy.md:8]: The size of batches should be unaffected. That's still totally controlled by the user's batch size setting :)
If you set concurrency to 1, batches are processed serially one at a time (which is how the system worked before the concurrency feature was added).

File: packages/wrangler/src/__tests__/publish.test.ts (hunk 10 above)
Comment (jbwcloudflare) [reply packages/wrangler/src/__tests__/publish.test.ts:7402]: 👍 Added!

File: packages/wrangler/src/__tests__/queues.test.ts (hunk 11 above)
Comment (jbwcloudflare) [reply packages/wrangler/src/__tests__/queues.test.ts:394]: I didn't add that flag, it wasn't needed for the`consumer add` command. For that command, the user can either pass a value (e.g. `--max-concurrency 5`) or omit it to select "no max".  I think I've got both of those cases covered in the tests.

File: packages/wrangler/src/queues/cli/commands/consumer/add.ts (hunk 12 above)
Comment (jbwcloudflare) [reply packages/wrangler/src/queues/cli/commands/consumer/add.ts]: Sure thing, added! Feel free to suggest alternative wording (anyone).

**Review by petebacondarwin — APPROVED** [review]
Thanks for the updates! LGTM!
